File: nodejs/sample-p12.js
const common = require("oci-common");
const identity = require("oci-identity");
const dbtools = require("oci-databasetools");
const secrets = require("oci-secrets");
const openssl = require('openssl-wrapper');
const fs = require('node:fs');
const oracledb = require('oracledb');
const { filestorage } = require("oci-sdk");


// compartment id
compartment_id = 'ocid1.tenancy.oc1..xxx'
// dbConnection_OCID
conn_id = 'ocid1.databasetoolsconnection.oc1.iad.xxx'
// oci profile
ociConfig = "~/.oci/config";
ociProfile = 'xxx';

//
// hopefully a much better way but this is what I found
//

function convertP12toPEM(pkcs12File,callback){
    fs.writeFile('ewallet.p12',Buffer.from( pkcs12File, 'base64'), { flag: 'w+' }, err => {});
    openssl.exec('pkcs12', {in: 'ewallet.p12', passout: `pass:${pkcs12Passwd}`,passin: `pass:${pkcs12Passwd}`},
    function(err, buffer) {
       console.log(err);
       results = buffer.toString();
       const lines = results.split("\n");
       filteredFile = '';
       isCert = false;
       const certs = [];
       lines.forEach(line => {
           if ( line.startsWith("-----BEGIN")) {
               isCert = true;
           }
           if ( isCert ){
               filteredFile = filteredFile + (filteredFile == ""? "": "\n") + line;  
           } 
           if ( line.startsWith("-----END")) {
               isCert = false;
               certs.push(filteredFile)
               filteredFile=""
           }
       })
       fs.writeFile('ewallet.pem',certs[0] + "\n" +
                                  certs[3]  + "\n" +
                                  certs[1]  + "\n" + 
                                  certs[2]  + "\n", { flag: 'w+' }, err => {});
       callback();
   });
}


(async () => {

    // JavaScript
    const provider = new common.ConfigFileAuthenticationDetailsProvider(
        ociConfig,
        ociProfile
    );

    const dbc = new dbtools.DatabaseToolsClient({authenticationDetailsProvider:provider});
    const secretsClient = new secrets.SecretsClient({authenticationDetailsProvider:provider});

    req = {
            databaseToolsConnectionId : conn_id
        }

    resp = await dbc.getDatabaseToolsConnection(req);

    dbUsername = resp.databaseToolsConnection.userName;
    connectionString = resp.databaseToolsConnection.connectionString;
    passwdOCID = resp.databaseToolsConnection.userPassword.secretId;
    pkcs12OCID = resp.databaseToolsConnection.keyStores[0].keyStoreContent.secretId;
    pkcs12PasswdOCID = resp.databaseToolsConnection.keyStores[0].keyStorePassword.secretId;

    passwdResp = await secretsClient.getSecretBundle({secretId:passwdOCID})
    passwd = Buffer.from(passwdResp.secretBundle.secretBundleContent.content.toString('utf8'), 'base64').toString('ascii');
    pkcs12PasswdResp = await secretsClient.getSecretBundle({secretId:pkcs12PasswdOCID})
    pkcs12Passwd = Buffer.from(pkcs12PasswdResp.secretBundle.secretBundleContent.content.toString('utf8'), 'base64').toString('ascii');

    pkcs12FileResp = await secretsClient.getSecretBundle({secretId:pkcs12OCID})


    pkcs12File = pkcs12FileResp.secretBundle.secretBundleContent.content;

    convertP12toPEM(pkcs12File, async function (){

        const connection = await oracledb.getConnection ({
            user            : dbUsername,
            password        : passwd,
            connectString   : connectionString,
            walletLocation  : ".",
            walletPassword  : pkcs12Passwd,
            retryCount      : 0,
            retryDelay      : 0 
        });
    
        const result = await connection.execute(
            `select * from dual`
        );
    
        console.log(result.rows);
        await connection.close();
    });


    
     console.log(pkcs12Passwd)

    })();

File: python/sample-p12.py
import oci
import base64
import os
import oracledb
from oci.config import from_file, validate_config
import time
from cryptography.hazmat.primitives.serialization \
        import pkcs12, Encoding, PrivateFormat, BestAvailableEncryption

# Wallet directory
walletDir = "wallet"
os.mkdir(walletDir)

 # Load an OCI config with the DEFAULT profile
oci_config = from_file(file_location="~/.oci/config", profile_name="KRISRICEIO")
# compartment_id
compartment_id = 'ocid1.tenancy.oc1..aaaaaaaaplcmdlf2lhfcqlkdid6hordp5zj5kb2gr3vhoeh4yghal2jqnqma'
# dbConnection_OCID
conn_id = 'ocid1.databasetoolsconnection.oc1.iad.amaaaaaakl7rxeyawfijlfsgfkx4fcp5sbzd3taa53wohxjpqfvionywp2pa'

# dbtoolsClient
dbtools_client = oci.database_tools.DatabaseToolsClient(config=oci_config)

# get Connection Details
dbt_resp = dbtools_client.get_database_tools_connection(conn_id)

# make a secret Client
secrets_client = oci.secrets.SecretsClient(oci_config)

def getSecret(ocid):
    secret_resp = secrets_client.get_secret_bundle(secret_id=ocid,
                                             stage="LATEST")
    decodedSecret = base64.b64decode(secret_resp.data.secret_bundle_content.content)
    return decodedSecret

# Associated DB's OCID
dbOCID = dbt_resp.data.related_resource.identifier
# DB Username
username = dbt_resp.data.user_name
# DB Connect String
connString = dbt_resp.data.connection_string
# DB User's PAssword
userPassword= getSecret(dbt_resp.data.user_password.secret_id).decode();
# P12/wallet Password
p12Password= getSecret(dbt_resp.data.key_stores[0].key_store_password.secret_id).decode()
# P12 File Contents
p12File = getSecret(dbt_resp.data.key_stores[0].key_store_content.secret_id)


#
# Convert the P12 to a .pem for use as p12 is not directly usable
#
def createPEMFromP12(p12Password,p12File):
    ## Write the p12 to the file system
    p12FileName= os.path.join(walletDir,'ewallet.p12')

    with open(p12FileName, 'wb') as p12_file:
        p12_file.write(p12File)

    ## extract .PEM from the P12
    pem_file_name =   os.path.join(walletDir,"ewallet.pem");

    pkcs12_data = open(p12FileName, "rb").read()
    result = pkcs12.load_key_and_certificates(pkcs12_data,
                                            p12Password.encode())
    private_key, certificate, additional_certificates = result
    encryptor = BestAvailableEncryption(p12Password.encode())

    with open(pem_file_name, "wb") as f:
        f.write(private_key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8,
                                        encryptor))
        f.write(certificate.public_bytes(Encoding.PEM))
        for cert in additional_certificates:
            f.write(cert.public_bytes(Encoding.PEM))

createPEMFromP12(p12Password,p12File)

connection = oracledb.connect(user=username, 
                              password=userPassword,  
                              dsn=connString, 
                              ssl_server_dn_match=False, 
                              wallet_password=p12Password,
                              wallet_location=walletDir)
with connection.cursor() as cursor:
    for result in cursor.execute("select * from dual"):
        print(result)